### ISSUE TITLE
Spurious linkage errors to GCC on Linux

### ISSUE THREAD

**Issue by carlocab** [Homebrew/brew#17303]
### `brew doctor` output

```shell
Your system is ready to brew.
```

### Verification

- [X] My "`brew doctor` output" above says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` twice and am still able to reproduce my issue.
- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### `brew config` output

```shell
HOMEBREW_VERSION: 4.3.0-18-gf4a9869
ORIGIN: https://github.com/Homebrew/brew
HEAD: f4a9869d27f5f3ece082e74da38a5a06a67a2637
Last commit: 3 hours ago
Core tap HEAD: c633ed746d07a54def764bdeaca1e224a2a9ad65
Core tap last commit: 4 minutes ago
Core tap JSON: 14 May 20:01 UTC
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_BOOTSNAP: set
HOMEBREW_CASK_OPTS: []
HOMEBREW_COLOR: set
HOMEBREW_CURL_PATH: /usr/bin/curl
HOMEBREW_DEVELOPER: set
HOMEBREW_FAIL_LOG_LINES: 150
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_GIT_EMAIL: 1589480+BrewTestBot@users.noreply.github.com
HOMEBREW_GIT_NAME: BrewTestBot
HOMEBREW_GIT_PATH: /usr/bin/git
HOMEBREW_LOGS: /__w/homebrew-core/homebrew-core/logs
HOMEBREW_MAKE_JOBS: 4
HOMEBREW_NO_AUTO_UPDATE: set
HOMEBREW_NO_COLOR: set
HOMEBREW_NO_EMOJI: set
HOMEBREW_NO_ENV_HINTS: set
HOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK: set
HOMEBREW_NO_INSTALL_FROM_API: set
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.1.4 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: quad-core 64-bit zen3
Clang: N/A
Git: 2.43.2 => /usr/bin/git
Curl: 7.81.0 => /usr/bin/curl
Kernel: Linux 6.5.0-1018-azure x86_64 GNU/Linux
OS: Ubuntu 22.04.4 LTS
Host glibc: 2.35
/usr/bin/gcc: 11.4.0
/usr/bin/ruby: N/A
glibc: N/A
gcc@11: N/A
gcc: N/A
xorg: N/A
```

### What were you trying to do (and why)?

Run `brew linkage --cached --test --strict` on a formula with an indirect GCC dependency, because it's part of the tests run by `test-bot`.

### What happened (include all command output)?

```
==> brew linkage --cached --test --strict sherlock
==> FAILED
Full linkage --cached --test --strict sherlock output
  Indirect dependencies with linkage:
    gcc
```

### What did you expect to happen?

No errors or output.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
# `sherlock` can be replaced with `sextractor` below if you don't want to wait for Homebrew/homebrew-core#171701
brew install sherlock
brew linkage --test --strict sherlock
```

**Comment by carlocab** [comment]
@p-linnane also mentioned something about an audit failure, but I don't see it in the CI log at Homebrew/homebrew-core#171701, so I'll let Patrick chime in on how to reproduce that.

**Comment by MikeMcQuaid** [comment]
Thanks @carlocab for opening!

Either `brew linkage` shouldn't complain about this or it should be fixed.

**Comment by bevanjkay** [comment]
https://github.com/Homebrew/homebrew-core/actions/runs/9088414062/job/24977982664?pr=171701#step:4:50

**Comment by p-linnane** [comment]
Repro: check out the PR, build and then:
```
==> brew linkage --cached sherlock
System libraries:
  /lib/x86_64-linux-gnu/libc.so.6
  /lib/x86_64-linux-gnu/libm.so.6
Homebrew libraries:
  /home/linuxbrew/.linuxbrew/opt/gcc/lib/gcc/current/libgcc_s.so.1 (gcc)
  /home/linuxbrew/.linuxbrew/opt/gcc/lib/gcc/current/libstdc++.so.6 (gcc)
Indirect dependencies with linkage:
  gcc
Dependencies with no linkage:
  certifi
```

**Comment by carlocab** [comment]
I was referring to this audit failure, which I can't find in the log:
```
Full audit --formula sherlock --online --new output
  sherlock
    * Formulae in homebrew/core should not have a Linux-only dependency on GCC.
  Error: 1 problem in 1 formula detected.
```

**Comment by Bo98** [comment]
That audit is correct. "Fixing" the linkage error by adding a dependency would fail that audit. The bug is that we shouldn't be flagging indirect (or undeclared) GCC linkage in `brew linkage` because of how the rpaths are set up.

**Comment by carlocab** [comment]
Right, ok, so the audit failure appears if you add a Linux-only dependency on GCC, which explains why I couldn't find it in the log.

**Comment by carlocab** [comment]
Should be fixed by #17306.
